### PR TITLE
Improve error message and print stack backtrace when failing to load models and canary_models

### PR DESCRIPTION
**Pull request by JansonJohnPZ**
With the fix, the error message will correctly show missing modules or indicate that a certain model does not exist.

Test Plan:

Test case 1: run a model with dependencies installed
```
(pytorch) PS D:\my git repo\benchmark> python run.py resnet50
Running eval method from resnet50 on cpu in eager mode with input batch size 32 and precision fp32.
CPU Total Wall Time: 2833.809 milliseconds
CPU Peak Memory:                0.6416 GB
```

Test case 2: run a model with missing dependencies
```
(pytorch) PS D:\my git repo\benchmark> python run.py yolov3
Traceback (most recent call last):
  File "D:\my git repo\benchmark\run.py", line 303, in <module>
    Model = load_model_by_name(args.model)
  File "D:\my git repo\benchmark\torchbenchmark\__init__.py", line 612, in load_model_by_name
    module = importlib.import_module(f'.models.{model_pkg}', package=__name__)
  File "C:\Users\temp\anaconda3\envs\pytorch\lib\importlib\__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1050, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1006, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 688, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 883, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "D:\my git repo\benchmark\torchbenchmark\models\yolov3\__init__.py", line 15, in <module>
    from .yolo_train import prepare_training_loop
  File "D:\my git repo\benchmark\torchbenchmark\models\yolov3\yolo_train.py", line 6, in <module>
    from torch.utils.tensorboard import SummaryWriter
  File "C:\Users\temp\anaconda3\envs\pytorch\lib\site-packages\torch\utils\tensorboard\__init__.py", line 1, in <module>
    import tensorboard
ModuleNotFoundError: No module named 'tensorboard'
```

Test case 3: run a canary_model with missing dependencies:

```
(pytorch) PS D:\my git repo\benchmark> python run.py gat
Warning: The model gat cannot be found at core set.
Traceback (most recent call last):
  File "D:\my git repo\benchmark\run.py", line 311, in <module>
    Model = load_canary_model_by_name(args.model)
  File "D:\my git repo\benchmark\torchbenchmark\__init__.py", line 625, in load_canary_model_by_name
    module = importlib.import_module(f'.canary_models.{model}', package=__name__)
  File "C:\Users\temp\anaconda3\envs\pytorch\lib\importlib\__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1050, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1006, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 688, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 883, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "D:\my git repo\benchmark\torchbenchmark\canary_models\gat\__init__.py", line 1, in <module>
    from torchbenchmark.util.framework.gnn.model_factory import GNNModel
  File "D:\my git repo\benchmark\torchbenchmark\util\framework\gnn\model_factory.py", line 6, in <module>
    from torch_geometric.nn import GAT, GCN, GraphSAGE
ModuleNotFoundError: No module named 'torch_geometric'
```

Test case 4: run a non-exist model

```
(pytorch) PS D:\my git repo\benchmark> python run.py abcde
Warning: The model abcde cannot be found at core set.
Error: The model abcde cannot be found at either core or canary model set.
```

### CODE REVIEW

File: torchbenchmark/__init__.py [hunk 1]
```diff
@@ -605,11 +608,9 @@ def load_model_by_name(model):
     assert len(models) == 1, f"Found more than one models {models} with the exact name: {model}"
```
Comment (xuzhao9) [inline torchbenchmark/__init__.py:608]: Should we also raise `ModelNotFoundError` if `model` is not found?

Comment (JansonJohnPZ) [reply torchbenchmark/__init__.py:609]: A warning message is added. This warning is triggered when the program tries to find a model in the canary list.
Test case 1: run a canary_model with missing dependencies:
```
(pytorch) PS D:\my git repo\benchmark> python run.py gat
Traceback (most recent call last):
  File "D:\my git repo\benchmark\run.py", line 304, in <module>
    Model = load_model_by_name(args.model)
  File "D:\my git repo\benchmark\torchbenchmark\__init__.py", line 607, in load_model_by_name
    raise RuntimeWarning(f"{model} is not found in the core list. Searching canary...")
RuntimeWarning: gat is not found in the core list. Searching canary...
Traceback (most recent call last):
  File "D:\my git repo\benchmark\run.py", line 312, in <module>
    Model = load_canary_model_by_name(args.model)
  File "D:\my git repo\benchmark\torchbenchmark\__init__.py", line 626, in load_canary_model_by_name
    module = importlib.import_module(f'.canary_models.{model}', package=__name__)
  File "C:\Users\temp\anaconda3\envs\pytorch\lib\importlib\__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1050, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1006, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 688, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 883, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "D:\my git repo\benchmark\torchbenchmark\canary_models\gat\__init__.py", line 1, in <module>
    from torchbenchmark.util.framework.gnn.model_factory import GNNModel
  File "D:\my git repo\benchmark\torchbenchmark\util\framework\gnn\model_factory.py", line 6, in <module>
    from torch_geometric.nn import GAT, GCN, GraphSAGE
ModuleNotFoundError: No module named 'torch_geometric'
```
Test case 2: run a non-exist model:
```
(pytorch) PS D:\my git repo\benchmark> python run.py abcde
Traceback (most recent call last):
  File "D:\my git repo\benchmark\run.py", line 304, in <module>
    Model = load_model_by_name(args.model)
  File "D:\my git repo\benchmark\torchbenchmark\__init__.py", line 607, in load_model_by_name
    raise RuntimeWarning(f"{model} is not found in the core list. Searching canary...")
RuntimeWarning: abcde is not found in the core list. Searching canary...
Traceback (most recent call last):
  File "D:\my git repo\benchmark\run.py", line 312, in <module>
    Model = load_canary_model_by_name(args.model)
  File "D:\my git repo\benchmark\torchbenchmark\__init__.py", line 625, in load_canary_model_by_name
    raise ModelNotFoundError(f"{model} is not found in the both core and canary model lists")
torchbenchmark.ModelNotFoundError: abcde is not found in the both core and canary model lists
```

File: torchbenchmark/__init__.py [hunk 2]
```diff
@@ -601,15 +604,14 @@ def load_model_by_name(model):
                     map(lambda y: os.path.basename(y), _list_model_paths()))
     models = list(models)
     if not models:
+        raise RuntimeWarning(f"{model} is not found in the core list. Searching canary...")
```
Comment (xuzhao9) [inline torchbenchmark/__init__.py:607]: Let's raise `ModelNotFoundError` here.

File: torchbenchmark/__init__.py [hunk 3]
```diff
@@ -619,12 +621,9 @@ def load_model_by_name(model):
...
 
 def load_canary_model_by_name(model: str):
-    assert _is_canary_model(model), f"Canary model {model} is not found in the {canary_model_dir} path."
-    try:
-        module = importlib.import_module(f'.canary_models.{model}', package=__name__)
-    except ModuleNotFoundError as e:
-        print(f"Warning: Could not find dependent module {e.name} for Model {model}, skip it. \n {e}")
-        return None
+    if not _is_canary_model(model):
+        raise ModelNotFoundError(f"{model} is not found in the both core and canary model lists")
```
Comment (xuzhao9) [inline torchbenchmark/__init__.py:625]:
```suggestion
        raise ModelNotFoundError(f"{model} is not found in the canary model list")
```

File: torchbenchmark/__init__.py (hunk 2 above)
Comment (xuzhao9) [inline torchbenchmark/__init__.py]:
```suggestion
        raise ModelNotFoundError(f"{model} is not found in the core list.")
```

File: torchbenchmark/__init__.py [hunk 4]
```diff
@@ -18,6 +18,9 @@
 from components._impl.tasks import base as base_task
 from components._impl.workers import subprocess_worker
 
+class ModelNotFoundError(Exception):
```
Comment (xuzhao9) [inline torchbenchmark/__init__.py:21]: Do not inherit `Exception`. Inherit from `RuntimeException`.

File: run.py [hunk 5]
```diff
@@ -16,9 +16,12 @@
 import torch
 import torch.profiler as profiler
 
+import traceback
+
 from torchbenchmark import load_canary_model_by_name, load_model_by_name
 from torchbenchmark.util.experiment.metrics import get_peak_memory
 
+
```
Comment (xuzhao9) [inline run.py:24]: Remove this empty line.

File: run.py [hunk 6]
```diff
@@ -295,12 +298,20 @@ def _validate_profile_options(profile_options: str):
...
     found = False
-    Model = load_model_by_name(args.model)
+
+    Model = None
+    try:
+        Model = load_model_by_name(args.model)
+    except ModuleNotFoundError:
+        traceback.print_exc()
+        exit(-1)
+    except RuntimeWarning:
```
Comment (xuzhao9) [inline run.py]:
```suggestion
    except ModelNotFoundError:
       try:
           Model = load_canary_model_by_name(args.model)
       except ModuleNotFoundError:
       except ModelNotFoundError:
```

**Comment by JansonJohnPZ** [comment]
Multiple changes were applied according to the above suggestions.

Here are new running results:

Test Plan:

Test case 1: run a model with dependencies installed
```
(pytorch) PS D:\my git repo\benchmark> python run.py resnet50
Running eval method from resnet50 on cpu in eager mode with input batch size 32 and precision fp32.
CPU Total Wall Time: 3112.708 milliseconds
CPU Peak Memory:                0.6504 GB
```

Test case 2: run a model with missing dependencies
```
(pytorch) PS D:\my git repo\benchmark> python run.py yolov3
Traceback (most recent call last):
  File "D:\my git repo\benchmark\run.py", line 302, in <module>
    Model = load_model_by_name(args.model)
  File "D:\my git repo\benchmark\torchbenchmark\__init__.py", line 613, in load_model_by_name
    module = importlib.import_module(f'.models.{model_pkg}', package=__name__)
  File "C:\Users\temp\anaconda3\envs\pytorch\lib\importlib\__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1050, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1006, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 688, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 883, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "D:\my git repo\benchmark\torchbenchmark\models\yolov3\__init__.py", line 15, in <module>
    from .yolo_train import prepare_training_loop
  File "D:\my git repo\benchmark\torchbenchmark\models\yolov3\yolo_train.py", line 6, in <module>
    from torch.utils.tensorboard import SummaryWriter
  File "C:\Users\temp\anaconda3\envs\pytorch\lib\site-packages\torch\utils\tensorboard\__init__.py", line 1, in <module>
    import tensorboard
ModuleNotFoundError: No module named 'tensorboard'
```

Test case 3: run a canary_model with missing dependencies
```
(pytorch) PS D:\my git repo\benchmark> python run.py gat
Traceback (most recent call last):
  File "D:\my git repo\benchmark\run.py", line 302, in <module>
    Model = load_model_by_name(args.model)
  File "D:\my git repo\benchmark\torchbenchmark\__init__.py", line 607, in load_model_by_name
    raise ModelNotFoundError(f"{model} is not found in the core model list.")
torchbenchmark.ModelNotFoundError: gat is not found in the core model list.
Traceback (most recent call last):
  File "D:\my git repo\benchmark\run.py", line 302, in <module>
    Model = load_model_by_name(args.model)
  File "D:\my git repo\benchmark\torchbenchmark\__init__.py", line 607, in load_model_by_name
    raise ModelNotFoundError(f"{model} is not found in the core model list.")
torchbenchmark.ModelNotFoundError: gat is not found in the core model list.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "D:\my git repo\benchmark\run.py", line 309, in <module>
    Model = load_canary_model_by_name(args.model)
  File "D:\my git repo\benchmark\torchbenchmark\__init__.py", line 626, in load_canary_model_by_name
    module = importlib.import_module(f'.canary_models.{model}', package=__name__)
  File "C:\Users\temp\anaconda3\envs\pytorch\lib\importlib\__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1050, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1006, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 688, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 883, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "D:\my git repo\benchmark\torchbenchmark\canary_models\gat\__init__.py", line 1, in <module>
    from torchbenchmark.util.framework.gnn.model_factory import GNNModel
  File "D:\my git repo\benchmark\torchbenchmark\util\framework\gnn\model_factory.py", line 6, in <module>
    from torch_geometric.nn import GAT, GCN, GraphSAGE
ModuleNotFoundError: No module named 'torch_geometric'
```

Test case 4: run a non-exist model
```
(pytorch) PS D:\my git repo\benchmark> python run.py abcde
Traceback (most recent call last):
  File "D:\my git repo\benchmark\run.py", line 302, in <module>
    Model = load_model_by_name(args.model)
  File "D:\my git repo\benchmark\torchbenchmark\__init__.py", line 607, in load_model_by_name
    raise ModelNotFoundError(f"{model} is not found in the core model list.")
torchbenchmark.ModelNotFoundError: abcde is not found in the core model list.
Traceback (most recent call last):
  File "D:\my git repo\benchmark\run.py", line 302, in <module>
    Model = load_model_by_name(args.model)
  File "D:\my git repo\benchmark\torchbenchmark\__init__.py", line 607, in load_model_by_name
    raise ModelNotFoundError(f"{model} is not found in the core model list.")
torchbenchmark.ModelNotFoundError: abcde is not found in the core model list.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "D:\my git repo\benchmark\run.py", line 309, in <module>
    Model = load_canary_model_by_name(args.model)
  File "D:\my git repo\benchmark\torchbenchmark\__init__.py", line 625, in load_canary_model_by_name
    raise ModelNotFoundError(f"{model} is not found in the canary model list.")
torchbenchmark.ModelNotFoundError: abcde is not found in the canary model list.
```

File: run.py [hunk 7]
```diff
@@ -295,12 +297,18 @@ def _validate_profile_options(profile_options: str):
...
-        Model = load_canary_model_by_name(args.model)
-        if not Model:
-            print(f"Unable to find model matching {args.model}.")
+
+    try:
+        Model = load_model_by_name(args.model)
+    except ModuleNotFoundError:
+        traceback.print_exc()
+        exit(-1)
+    except:
```
Comment (xuzhao9) [inline run.py:306]: Capture `ModelNotFoundError` here.
```suggestion
    except ModelNotFoundError:
```

File: run.py [hunk 8]
```diff
@@ -295,12 +297,18 @@ def _validate_profile_options(profile_options: str):
...
+        traceback.print_exc()
```
Comment (xuzhao9) [inline run.py:307]: No need to print stack backtrace if it is model not found.

File: torchbenchmark/__init__.py [hunk 9]
```diff
@@ -601,15 +604,14 @@ def load_model_by_name(model):
                     map(lambda y: os.path.basename(y), _list_model_paths()))
     models = list(models)
     if not models:
+        raise ModelNotFoundError(f"{model} is not found in the core model list.")
         return None
```
Comment (xuzhao9) [inline torchbenchmark/__init__.py:608]: No need to `return None` if you raise exception here. Remove this line.

**Comment by JansonJohnPZ** [comment]
Redundant lines are removed, and the running result remains unchanged.

**Comment by JansonJohnPZ** [comment]
Every comment is marked as resolved now.

File: run.py (hunk 7 above)
Comment (xuzhao9) [reply run.py:305]: cc @JansonJohnPZ  Can you address my comment here?

**Comment by xuzhao9** [comment]
@JansonJohnPZ There is one remaining

File: run.py (hunk 7 above)
Comment (JansonJohnPZ) [reply run.py]: This line is added:
```
except ModelNotFoundError
```

I found the error is caught in run.py. To print the traceback message, I restore the following line:
```
traceback.print_exc()
```

File: run.py (hunk 8 above)
Comment (xuzhao9) [reply run.py:307]: cc @JansonJohnPZ Can you remove the stacktrace print when we met "ModelNotFoundError"?

**Comment by xuzhao9** [comment]
Added 2 more comments, can you address?

File: run.py [hunk 10]
```diff
@@ -295,12 +297,18 @@ def _validate_profile_options(profile_options: str):
...
+    except ModuleNotFoundError:
+        traceback.print_exc()
+        exit(-1)
+    except ModelNotFoundError:
+        traceback.print_exc()
+        try:
+            Model = load_canary_model_by_name(args.model)
+        except ModuleNotFoundError:
+            traceback.print_exc()
             exit(-1)
```
[the remark this answers is not in the document]
Comment (xuzhao9) [reply run.py:312]: cc @JansonJohnPZ Can you address this as well?

[the remark this answers is not in the document]
Comment (JansonJohnPZ) [reply run.py:312]: Sure, I will address them.

**Comment by JansonJohnPZ** [comment]
When not finding the model, there will not be more traceback print.

Test case 1, run a canary_model with missing dependencies:
```
(pytorch) PS D:\my git repo\benchmark> python run.py gat
Warning: The model gat cannot be found at core set.
Traceback (most recent call last):
  File "D:\my git repo\benchmark\run.py", line 311, in <module>
    Model = load_canary_model_by_name(args.model)
  File "D:\my git repo\benchmark\torchbenchmark\__init__.py", line 625, in load_canary_model_by_name
    module = importlib.import_module(f'.canary_models.{model}', package=__name__)
  File "C:\Users\temp\anaconda3\envs\pytorch\lib\importlib\__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1050, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1006, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 688, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 883, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "D:\my git repo\benchmark\torchbenchmark\canary_models\gat\__init__.py", line 1, in <module>
    from torchbenchmark.util.framework.gnn.model_factory import GNNModel
  File "D:\my git repo\benchmark\torchbenchmark\util\framework\gnn\model_factory.py", line 6, in <module>
    from torch_geometric.nn import GAT, GCN, GraphSAGE
ModuleNotFoundError: No module named 'torch_geometric'
```

Test case 2, run a non-exist model:
```
(pytorch) PS D:\my git repo\benchmark> python run.py abcde
Warning: The model abcde cannot be found at core set.
Error: The model abcde cannot be found at either core or canary model set.
```

**Comment by xuzhao9** [comment]
Looks good to me now. Can you please update the output of 4 test cases in the PR body?

**Comment by JansonJohnPZ** [comment]
> 

Update the output of 4 test cases in the PR body is done.

**Comment by xuzhao9** [comment]
Looks good, thanks for your contribution to Torchbench!

**Review by xuzhao9 — APPROVED** [review]
LGTM!

**Pull request closed without merging**